### DISCUSSION TITLE
Performance issues with large amounts of contacts

### DISCUSSION THREAD

**Discussion by peter1745** [jrouwe/JoltPhysics#416] [Q&A] [answered]
Hello! Just want to say that I decided to make this "issue" a discussion because I'm not sure it's necessarily an actual issue, and it might be more useful as a discussion instead of an issue.

I'm currently wrapping up replacing PhysX with Jolt in a game engine I'm working on for work, and as a result I'm trying to optimize parts of Jolt that I didn't consider before.
As a result I'm currently stress testing Jolt, and while it's been performing much better than PhysX I did recently come across a pretty serious performance issue when there's a large number of contacts being generated in the scene.

I decided to setup a scene where we have a total of 5625 dynamic bodies spawned in as four pillars, so ~1406 bodies per pillar, each body has a basic box shape attached.

Initially the performance is fine, but as the bodies start colliding with each other after hitting the ground there's noticeable performance loss, which isn't really too surprising, it's a *lot* of a bodies.

From my testing the performance starts to suffer after there's around 4000 contacts being generated every "frame", and it gets even worse as more contacts start to be generated, I think the max number of contacts I ever encountered with this particular test was ~20,000.

I will say that the performance gets much better as bodies starts being deactivated, but I think there's so many small forces constantly being applied (due to all bodies piling on top of each other) that it takes a really long time for the bodies to be deactivated.

I decided to use Visual Studios Performance Profiler to capture snapshots as more contacts were generated, the thing that I found strange was that Visual Studios profiling results shows that `ContactConstraintManager::SolveVelocityConstraints` is the slowest function when there's a lot of contacts, taking up roughly 28% of the total CPU time (GitHub doesn't let me send .diagsession files so I can't attach the profiling results).

![image](https://user-images.githubusercontent.com/7275084/213836442-b4bd4b48-6900-4652-b6cf-7735c89066b3.png)

Specifically it seems like `sSolveVelocityConstraint` between two dynamic bodies is the worst "offender" (sadly Visual Studios profiler doesn't take snapshots of that for whatever reason).
![image](https://user-images.githubusercontent.com/7275084/213836759-18fe2740-8102-4d9d-b4f9-41a386f992ca.png)

My current working theory is that it's just that solving the velocity constraints is a relatively heavy operation, and that it's registering as a hot path because it's called so many times due to there being a lot of bodies.

I'm not really sure if it's possible to optimize it further, or if it'd be worth the effort of doing so, as I said I'm stress testing Jolt here so it's not a realistic use case by far, and if it's not worth the effort then that's understandable, especially since Jolt has been holding up much better compared to our previous PhysX implementation.

I really just opened this discussion because I found this interesting and was curious about it so if I'm completely on the wrong track with my thinking then I guess I might learn something at least, I've been learning quite a lot just from using Jolt and looking into how it works so it's been a really good experience so far 🙂

**Comment by jrouwe** [comment, marked as answer]
Hello,

Have you looked at a profile per thread? You can e.g. make one by calling JPH_PROFILE_DUMP.

What I think is happening is that you're probably generating a very big pile of bodies, where everything (indirectly) touches everything. This means that the bodies will form 1 big simulation island and since Jolt distributes the simulation across multiple threads based on these simulation islands you essentially fall back to a single threaded simulation. E.g. this capture comes from the FunnelTest in the Samples application when everything has nearly come to rest:

![image](https://user-images.githubusercontent.com/1621693/213861340-7d5a30a2-4ec1-46da-ba61-4a1a8f1c587d.png)

You can see that SolveVelocityConstraints became a big bottleneck. Probably if you put your pillars further apart so that bodies from 1 pillar cannot touch bodies from the next pillar, you'll see that there will be 4 SolveVelocityConstraints jobs in parallel meaning the total throughput of the system is 4x as big (even though the amount of work is roughly the same).

This is a weak point of Jolt and acceptable for me as I don't have scenes where everything ends up in 1 pile, but I do intend to split these simulation islands up further to allow at least some parallelism in these cases somewhere in the future.

**Reply by peter1745** [reply]
Thanks for the answer, that explains what I've been seeing perfectly because when it was at its worst was when I tried having all cubes be part of a single pillar.
It's not really something I'm too concerned about to be honest, more of a curiosity, and despite that Jolt performed very well even when there was around 2000 bodies in a fairly small area.

Anyway thanks for taking the time to explain why the performance was suffering in my test case, I really appreciate it!
